embassy-stm32/src/usart/mod.rs

```diff
@@ -536,21 +946,36 @@ unsafe fn clear_interrupt_flags(r: Regs, sr: regs::Isr) {
 }
 
 pub(crate) mod sealed {
+    use embassy_sync::waitqueue::AtomicWaker;
+
     use super::*;
 
+    pub struct State {
+        pub rx_waker: AtomicWaker,
```

Comment:
Have you considered using a [Signal](https://docs.embassy.dev/embassy-sync/git/default/signal/struct.Signal.html) to share state between interrupt and async function? I'm not sure this is used like this anywhere else, but I don't see why this wouldn't work here. This way your checks in `on_interrupt` could send the event trigger (specific error, idle...), to your async function, where you don't need to check the registers again. You could also immediately clear any interrupt flags in the interrupt.

Reply:
In an intermediate implementation, I used an AtomicU32 to store a copy of the status register, which should be roughly equivalent to the Signal. I asked on the chat if that was the best way to share state between the interrupt handler and the future and I got the answer to keep the status register flags set but disable the corresponding interrupt.
This saves 4 bytes in the state for each Uart at the expanse of an extra read of the status register. I guess there may be situations were we don't care about the extra memory consumption but I think it is a reasonable approach.